packages/custom_lint/lib/src/package_utils.dart

```diff
@@ -115,14 +123,74 @@ Future<PackageConfig?> tryParsePackageConfig(Directory directory) async {
   }
 }
 
+/// Finds the project directory associated with an analysis context root
+///
+/// This is a folder that contains both a `pubspec.yaml` and a `.dart_tool/package_config.json` file.
+/// It is either alongside the analysis_options.yaml file, or in a parent directory.
+Directory findProjectDirectory(
+  Directory directory, {
+  Directory? original,
+  bool missingPackageConfigOkay = false,
+}) {
+  final packageConfigFile = directory.packageConfig;
+  final pubspecFile = directory.pubspec;
+  if (packageConfigFile.existsSync() && pubspecFile.existsSync()) {
+    return directory;
+  }
+  if (pubspecFile.existsSync()) {
+    if (missingPackageConfigOkay) {
+      return directory;
+    }
+    throw PackageConfigNotFoundError._(directory.path);
+  }
+  if (directory.parent.uri == directory.uri) {
+    throw FindProjectError._(original?.path ?? directory.path);
```

Comment:
I'd rather make the return value nullable.

Reply:
Added a tryFindProjectDirectory

Reply:
That's not quite what I meant. I don't think we need a variant that throws. If we do want to throw, I'd move the exception to the place which calls `findProductDirectory` instead.